### PR TITLE
Support modals

### PR DESCRIPTION
**Pull request by ghost**
Adds support for sending modal components and receiving modal submit interactions.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -66,7 +66,7 @@ declare namespace Eris {
   type GuildIntegrationExpireBehavior = Constants["GuildIntegrationExpireBehavior"][keyof Constants["GuildIntegrationExpireBehavior"]];
 
   // Message
-  type ActionRowComponents = Button | SelectMenu;
+  type ActionRowComponents = Button | SelectMenu | TextInput;
```
Comment (coolcalcacol) [inline index.d.ts]: Is this the best way to do it? as Text inputs are exclusive to modals?

File: index.d.ts [hunk 2]
```diff
@@ -2933,12 +2951,18 @@ declare namespace Eris {
     custom_id: string;
   }
 
+
```
Comment (coolcalcacol) [inline index.d.ts]: random newline?

File: index.d.ts (hunk 1 above)
Comment (ghost) [reply index.d.ts]: Text input is a component type according to the docs so...?

**Comment by bsian03** [comment]
I already had a draft version of this but didn't push it as there are other interaction related stuff I wanted to merge first, since there will be conflicts in various different places

**Comment by DonovanDMC** [comment]
we're bordering on 4 or 5 conflicts at this point, with another one on the horizon (attachment option type)

**Comment by bsian03** [comment]
There's also a reason why I organised the typings file the way I did, please follow the structure in that file please

File: lib/structures/CommandInteraction.js [hunk 3]
```diff
@@ -237,6 +237,27 @@ class CommandInteraction extends Interaction {
         }, file).then(() => this.update());
     }
 
+    /**
```
Comment (davidffa) [inline lib/structures/CommandInteraction.js:245]: Missing method description.

File: lib/structures/ComponentInteraction.js [hunk 4]
```diff
@@ -170,6 +170,27 @@ class ComponentInteraction extends Interaction {
         }, file).then(() => this.update());
     }
 
+    /**
```
Comment (davidffa) [inline lib/structures/ComponentInteraction.js:178]: Missing description here too.

**Review by davidffa — CHANGES_REQUESTED** [review]
Missing JSDoc description on createModal function.

File: esm.mjs [hunk 5]
```diff
@@ -59,4 +60,4 @@ export const {
   VoiceConnection,
   VoiceConnectionManager,
   VoiceState
-} = Eris;
+} = Eris;
```
Comment (bsian03) [inline esm.mjs]: EOF newline

File: index.d.ts [hunk 6]
```diff
@@ -1564,6 +1584,11 @@ declare namespace Eris {
       ACTION_ROW:  1;
       BUTTON:      2;
       SELECT_MENU: 3;
+      TEXT_INPUT:  4;
+    };
+    TextInputStyles: {
```
Comment (bsian03) [inline index.d.ts]: alphabetical order

File: index.d.ts [hunk 7]
```diff
@@ -2975,6 +3004,35 @@ declare namespace Eris {
     acknowledge(choices: ApplicationCommandOptionChoice[]): Promise<void>;
     result(choices: ApplicationCommandOptionChoice[]): Promise<void>;
   }
+
+  interface ModalSubmitInteractionDataComponents {
```
Comment (bsian03) [inline index.d.ts]: incorrect location

File: index.d.ts [hunk 8]
```diff
@@ -2975,6 +3004,35 @@ declare namespace Eris {
...
+    components: (Pick<TextInput, "custom_id" | "type"> & { value: string })[];
+    type: Constants["ComponentTypes"]["ACTION_ROW"];
+  }
+  interface ModalSubmitInteractionData {
```
Comment (bsian03) [inline index.d.ts]: incorrect location

File: index.d.ts [hunk 9]
```diff
@@ -2975,6 +3004,35 @@ declare namespace Eris {
...
+    custom_id: string;
+    components: ModalSubmitInteractionDataComponents[];
+  }
+
+  export class ModalSubmitInteraction<T extends PossiblyUncachedTextable = TextableChannel> extends Interaction {
```
Comment (bsian03) [inline index.d.ts:3078]: alphabetical order

File: lib/Constants.js [hunk 10]
```diff
@@ -108,6 +108,11 @@ module.exports.ButtonStyles = {
     LINK:      5
 };
 
+module.exports.TextInputStyles = {
```
Comment (bsian03) [inline lib/Constants.js]: alphabetical order

File: lib/structures/Interaction.js [hunk 11]
```diff
@@ -4,7 +4,7 @@ const Base = require("./Base");
 const {InteractionTypes} = require("../Constants");
 
 /**
-* Represents an interaction. You also probably want to look at PingInteraction, CommandInteraction, ComponentInteraction, AutocompleteInteraction, and UnknownInteraction.
+* Represents an interaction. You also probably want to look at PingInteraction, CommandInteraction, ComponentInteraction, AutocompleteInteraction, ModalSubmitInteraction, and UnknownInteraction.
 * @prop {Boolean} acknowledged Whether or not the interaction has been acknowledged
 * @prop {String} applicationID The ID of the interaction's application
 * @prop {String} id The ID of the interaction
```
Comment (bsian03) [inline lib/structures/Interaction.js:10]: missing type description on L12

File: lib/structures/ModalSubmitInteraction.js [hunk 12]
```diff
@@ -0,0 +1,400 @@
+"use strict";
+
+const Interaction = require("./Interaction");
+const Member = require("./Member");
+const {InteractionResponseTypes} = require("../Constants");
+
+/**
+* Represents a message component interaction. See Interaction for more properties.
+* @extends Interaction
+* @prop {PrivateChannel | TextChannel | NewsChannel} channel The channel the interaction was created in. Can be partial with only the id if the channel is not cached.
```
Comment (bsian03) [inline lib/structures/ModalSubmitInteraction.js:10]: is this uncached documentation wording consistent with the same prop in message? it definitely doesn't match event docs

**Review by bsian03 — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 9 above)
Comment (ghost) [reply index.d.ts:3078]: The order is completely messed if we're considering alphabetical order. AutocompleteInteraction is after ComponentInteraction. Do you want me to re-arrange them?

**Comment by mremoji0808** [comment]
image
image
1. Should data.component_type be present in `ModalSubmitInteraction` class? 
2. In `index.t.s` ComponentInteractionTextInputData is missing. (if only component_type should be there else can be remove)
image

**Comment by ghost** [comment]
What? Can you provide links to the snippets? I'm not able to find `ComponentInteractionInputTextData` interface either.

**Review by bsian03 — DISMISSED** [review]
Fix merge conflict and sort all prop names alphabetically, also missing props from https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-response-object-modal

File: index.d.ts [hunk 13]
```diff
@@ -2998,6 +3036,23 @@ declare namespace Eris {
...
+  export class AutocompleteInteraction<T extends PossiblyUncachedTextable = TextableChannel> extends Interaction {
+    appPermissions?: Permission;
+    channel: T;
+    data: {
+      id: string;
+      name: string;
+      type: Constants["ApplicationCommandTypes"]["CHAT_INPUT"];
+      target_id?: string;
+      options: InteractionDataOptions[];
+    };
```
Comment (bsian03) [inline index.d.ts:3048]: Interface this

File: lib/structures/ModalSubmitInteraction.js [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+                const guild = this._client.guilds.get(info.guild_id);
+                this.member = new Member(info.member, guild, this._client);
+            }
+        }
+
+        if(info.user !== undefined) {
+            this.user = this._client.users.update(info.user, client);
+        }
+
+    }
```
Comment (bsian03) [inline lib/structures/ModalSubmitInteraction.js:46]: Unnecessary/inconsistent newlines

File: lib/structures/ModalSubmitInteraction.js [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+* @prop {PrivateChannel | TextChannel | NewsChannel} channel The channel the interaction was created in. Can be partial with only the id if the channel is not cached.
+* @prop {Object} data The data attached to the interaction
+* @prop {String} data.custom_id The ID of the Modal
+* @prop {Array<Object>} data.components The values submitted by the user
+* @prop {String?} guildID The ID of the guild in which the interaction was created
+* @prop {Member?} member The member who triggered the interaction (This is only sent when the interaction is invoked within a guild)
+* @prop {User?} user The user who triggered the interaction (This is only sent when the interaction is invoked within a dm)
+*/
+class ModalSubmitInteraction extends Interaction {
+    constructor(info, client) {
```
Comment (bsian03) [inline lib/structures/ModalSubmitInteraction.js:19]: Should be called `data` instead of `info`

File: lib/structures/ModalSubmitInteraction.js [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+            this.guildID = info.guild_id;
+        }
+
+        if(info.member !== undefined) {
+            if(this.channel.guild) {
+                info.member.id = info.member.user.id;
+                this.member = this.channel.guild.members.update(info.member, this.channel.guild);
+            } else {
+                const guild = this._client.guilds.get(info.guild_id);
+                this.member = new Member(info.member, guild, this._client);
```
Comment (bsian03) [inline lib/structures/ModalSubmitInteraction.js:38]: Upsert into cache

File: index.d.ts [hunk 17]
```diff
@@ -1351,6 +1371,16 @@ declare namespace Eris {
     locked: boolean;
   }
 
+  // Modals
+  interface ModalSubmitInteractionDataComponents {
+    components: (Pick<TextInput, "custom_id" | "type"> & { value: string })[];
```
Comment (bsian03) [inline index.d.ts:1376]: Interface this

**Review by bsian03 — CHANGES_REQUESTED** [review]

**Comment by HeadTriXz** [comment]
`createModal` type is missing on `ComponentInteraction`.

File: lib/Client.js [hunk 18]
```diff
@@ -733,22 +733,26 @@ class Client extends EventEmitter {
...
-    * @arg {Number} [options.data.components[].style] The style of the component (type 2 only) - If 0-4, `custom_id` is required; if 5, `url` is required
+    * @arg {String} [options.data.components[].placeholder] The placeholder text for the component when no option is selected or the text input is empty (type 3 and 4)
+    * @arg {Number} [options.data.components[].style] The style of the component (type 2 and type 4 only) - For buttons, if 0-4, `custom_id` is required; if 5, `url` is required, for text inputs 1-2
     * @arg {Number} options.data.components[].type The type of component - If 1, it is a collection and a `components` array (nested) is required; if 2, it is a button; if 3, it is a select menu
     * @arg {String} [options.data.components[].url] The URL that the component should open for users (type 2 style 5 only)
+    * @arg {Boolean} [options.data.components[].required] Whether this component is required to be filled, default false (type 4)
+    * @arg {String} [options.data.components[].value] A a pre-filled value for this component, max 4000 characters (type 4)
```
Comment (DonovanDMC) [inline lib/Client.js:755]: We should remove this mess and just replace it with a link to Discord's docs, like we've done for `embed`/`embeds` and the like

**Comment by DonovanDMC** [comment]
Are you planning on reopening this, or are you dropping the pr?

**Comment by ghost** [comment]
Dropping

**Comment by ghost** [comment]
I won't delete the branch for now, if anyone wants to use the code.

**Comment by DonovanDMC** [comment]
Recreated as #1404

**Comment by carloslimaborges** [comment]
I can carry on if you guys prefer. I've been using his fork in one of my bots. Btw thank you so much @Linker-123

**Comment by carloslimaborges** [comment]
nvm, looks like @DonovanDMC already did :)

**Pull request closed without merging**